Carbon/Functions/Resolve-CPrivateKeyPath.ps1

```diff
@@ -0,0 +1,205 @@
+
+function Resolve-CPrivateKeyPath
+{
+    <#
+    .SYNOPSIS
+    Finds the path to a certificate's private key.
+
+    .DESCRIPTION
+    The `Resolve-CPrivateKeyPath` function finds the path to a certificate's private key. Pipe the certificate object to
+    the function (or pass one or more to the `Certificate` parameter). The function searches all the directories where
+    keys are stored, [which are documented by
+    Microsoft](https://learn.microsoft.com/en-us/windows/win32/seccng/key-storage-and-retrieval).
+
+    If the certificate doesn't have a private key, have access to the private key, or no private key file exists, the
+    function writes an error and returns nothing for that certificate.
+
+    Returns the path to the private key as a string.
+
+    .LINK
+    https://learn.microsoft.com/en-us/windows/win32/seccng/key-storage-and-retrieval
+
+    .EXAMPLE
+    $cert | Resolve-CPrivateKeyPath
+
+    Demonstrates that you can pipe X509Certificate2 objects to this function.
+
+    .EXAMPLE
+    Resolve-CPrivateKeyPath -Certificate $cert
+
+    Demonstrates that you pass an X509Certificate2 object to the `Certificate` parameter.
+    #>
+    [CmdletBinding()]
+    [OutputType([String])]
+    param(
+        # The certificate whose private key path to get. Must have a private key and that private key must be accessible
+        # by the current user.
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [Security.Cryptography.X509Certificates.X509Certificate2[]] $Certificate
+    )
+
+    begin
+    {
+        Set-StrictMode -Version 'Latest'
+        Use-CallerPreference -Cmdlet $PSCmdlet -Session $ExecutionContext.SessionState
+
+        $searchPaths =
+            & {
+                $appData = [Environment]::GetFolderPath('ApplicationData')
+                if ($appData)
+                {
+                    if ($IsWindows)
+                    {
+                        $sid = [Security.Principal.WindowsIdentity]::GetCurrent().User
+                        $sidString = $sid.ToString()
+
+                        # CSP user private
+                        Join-Path -Path $appData -ChildPath "Microsoft\Crypto\RSA\${sidString}"
+                        Join-Path -Path $appData -ChildPath "Microsoft\Crypto\DSS\${sidString}"

```

Comment:
I've started writing paths with forward slashes in PowerShell since forward slashes work across all platforms but back slashes are Windows-only. Just in case a future version of PowerShell decides to only support forward slashes for paths.

Reply:
As long as you use `Join-Path` to create a path, you don't have to worry about which slash you're using: `Join-Path` changes them to the appropriate slash for the current OS:

```
# On Windows
> Join-Path -Path 'C:/Windows/' -ChildPath 'ServiceProfiles/LocalService' 
C:\Windows\ServiceProfiles\LocalService

# On Ubuntu
> Join-Path -Path 'C:\Windows\' -ChildPath 'ServiceProfiles\LocalService'
C:/Windows/ServiceProfiles/LocalService
```